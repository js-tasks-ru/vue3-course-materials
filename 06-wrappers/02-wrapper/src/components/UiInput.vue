<template>
  <div
    class="input-group"
    :class="{
      'input-group_icon': $slots['left-icon'],
      'input-group_icon-left': $slots['left-icon'],
    }"
  >
    <div v-if="$slots['left-icon']" class="input-group__icon">
      <slot name="left-icon" />
    </div>
    <!--
      v-bind="$attrs" позволяет пробросить все переданные параметры и обработчики событий,
      которые не часть интерфейса компоненте (не описаны в props / emits),
      пробросить на определённый элемент внутри компонента.
      По умолчанию они наследуются корневым узлом.
    -->
    <input ref="input" v-model="modelProxy" v-bind="$attrs" class="form-control" :class="{ 'form-control_rounded': rounded }" />
    <!-- Или вместо modelProxy: -->
    <!-- :value="modelValue"-->
    <!-- @input="$emit('update:modelValue', $event.target.value)"-->
  </div>
</template>

<script>
export default {
  name: 'UiInput',

  // Отключаем стандартное наследование атрибутов
  // Иначе все атрибуты, которые мы вручную проксируем на input, будут дублироваться и на корневом div-е
  inheritAttrs: false,

  props: {
    modelValue: {
      type: String,
    },

    rounded: {
      type: Boolean,
    }
  },

  emits: {
    'update:modelValue': null,
  },

  // По умолчанию (кроме SFC Setup) все свойства компонента публичны.
  // Можно явно описать, какие публичны. Тогда остальные будут приватные.
  expose: ['focus'],

  computed: {
    // Не работает. $slots - не реактивный
    // hasIcon() {
    //   return !!this.$slots['left-icon'];
    // },

    // Проксирование модели, позволяет v-model компонента напрямую связать с v-model оборачиваемого элемента
    // Особенно полезно с элементами со сложной моделью, например, checkbox
    modelProxy: {
      get() {
        return this.modelValue;
      },

      set(value) {
        this.$emit('update:modelValue', value);
      },
    },
  },

  methods: {
    // В редких случаях методы могут быть публичным интерфейсом компонента
    focus() {
      this.$refs['input'].focus();
    },
  },
};
</script>

<style scoped>
.form-control {
  padding: 12px 16px;
  height: 52px;
  border-radius: 8px;
  border: 2px solid var(--blue-light);
  font-family: 'Nunito', sans-serif;
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: var(--body-color);
  transition: 0.2s border-color;
  background-color: var(--white);
  outline: none;
  box-shadow: none;
}

.form-control::placeholder {
  font-weight: 400;
  color: var(--blue-2);
}

.form-control:focus {
  border-color: var(--blue);
}

textarea.form-control {
  width: 100%;
  min-height: 211px;
}

.form-control.form-control_rounded {
  border-radius: 26px;
}

.form-control.form-control_sm.form-control_rounded {
  border-radius: 22px;
}

.form-control.form-control_sm {
  padding: 8px 16px;
  height: 44px;
  border-radius: 4px;
}

.input-group {
  position: relative;
}

.input-group .form-control {
  width: 100%;
}

.input-group.input-group_icon .form-control {
  padding-left: 50px;
}

.input-group.input-group_icon .input-group__icon {
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
}

.input-group.input-group_icon.input-group_icon-left .input-group__icon {
  left: 16px;
}

.input-group.input-group_icon.input-group_icon-right .input-group__icon {
  right: 16px;
}
</style>
