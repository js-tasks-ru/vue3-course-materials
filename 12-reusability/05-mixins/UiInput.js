import { defineComponent } from './vendor/vue.esm-browser.js';
import { modelProxyMixin } from './mixins/modelProxyMixin.js';

export default defineComponent({
  name: 'UiInput',

  // Локальное подключение миксинов
  mixins: [modelProxyMixin],

  props: ['modelValue'],

  emits: ['update:modelValue'],

  template: `<input v-model="modelValueProxy" />`,
});
