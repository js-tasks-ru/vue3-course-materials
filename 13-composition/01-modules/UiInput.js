import { defineComponent } from './vendor/vue.esm-browser.js';

export default defineComponent({
  name: 'UiInput',

  props: ['modelValue'],

  emits: ['update:modelValue'],

  computed: {
    modelValueProxy: {
      get() {
        return this.modelValue;
      },

      set(value) {
        this.$emit(`update:modelValue`, value);
      },
    },
  },

  template: `<input v-model="modelValueProxy" />`,
});
