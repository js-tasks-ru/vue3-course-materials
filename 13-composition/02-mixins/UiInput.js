import { defineComponent } from './vendor/vue.esm-browser.js';
import { createModelProxy } from './utils/createModelProxy.js';

export default defineComponent({
  name: 'UiInput',

  props: ['modelValue'],

  emits: ['update:modelValue'],

  computed: {
    modelValueProxy: createModelProxy(),
  },

  template: `<input v-model="modelValueProxy" />`,
});
