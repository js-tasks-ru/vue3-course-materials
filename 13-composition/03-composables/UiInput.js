import { computed, toRefs } from './vendor/vue.esm-browser.js';
import { defineComponent } from './vendor/vue.esm-browser.js';
import { useModelProxy } from './composables/useModelProxy.js';

export default defineComponent({
  name: 'UiInput',

  props: ['modelValue'],

  emits: ['update:modelValue'],

  setup(props, { emit }) {
    const { modelValue } = toRefs(props);
    const modelValueProxy = computed({
      get() {
        return modelValue.value;
      },

      set(value) {
        emit('update:modelValue', value);
      },
    });

    return {
      modelValueProxy,
    };
  },

  template: `<input v-model="modelValueProxy" />`,
});
