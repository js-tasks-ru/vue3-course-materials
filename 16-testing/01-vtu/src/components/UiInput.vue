<template>
  <div
    class="input-group"
    :class="{
      'input-group_icon': $slots['left-icon'],
      'input-group_icon-left': $slots['left-icon'],
    }"
  >
    <div v-if="$slots['left-icon']" class="input-group__icon">
      <slot name="left-icon" />
    </div>
    <input
      ref="input"
      v-model="customModel"
      v-bind="$attrs"
      class="form-control"
      :class="{
        'form-control_rounded': rounded,
        'form-control_sm': small,
      }"
    />
    <!-- Или вместо customModel: -->
    <!-- :value="modelValue"-->
    <!-- @input="$emit('update:modelValue', $event.target.value)"-->
  </div>
</template>

<script>
export default {
  name: 'UiInput',

  inheritAttrs: false,

  props: {
    modelValue: {
      type: String,
    },

    rounded: Boolean,
    small: Boolean,
  },

  emits: {
    'update:modelValue': null,
  },

  computed: {
    // Не работает. $slots - не реактивный
    // hasIcon() {
    //   return !!this.$slots['left-icon'];
    // },

    customModel: {
      get() {
        return this.modelValue;
      },

      set(value) {
        this.$emit('update:modelValue', value);
      },
    },
  },

  methods: {
    focus() {
      this.$refs['input'].focus();
    },
  },
};
</script>

<style scoped>
.form-control {
  padding: 12px 16px;
  height: 52px;
  border-radius: 8px;
  border: 2px solid var(--blue-light);
  font-family: 'Nunito', sans-serif;
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: var(--body-color);
  transition: 0.2s border-color;
  background-color: var(--white);
  outline: none;
  box-shadow: none;
}

.form-control::placeholder {
  font-weight: 400;
  color: var(--blue-2);
}

.form-control:focus {
  border-color: var(--blue);
}

textarea.form-control {
  width: 100%;
  min-height: 211px;
}

.form-control.form-control_rounded {
  border-radius: 26px;
}

.form-control.form-control_sm.form-control_rounded {
  border-radius: 22px;
}

.form-control.form-control_sm {
  padding: 8px 16px;
  height: 44px;
  border-radius: 4px;
}

.input-group {
  position: relative;
}

.input-group .form-control {
  width: 100%;
}

.input-group.input-group_icon .form-control {
  padding-left: 50px;
}

.input-group.input-group_icon .input-group__icon {
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
}

.input-group.input-group_icon.input-group_icon-left .input-group__icon {
  left: 16px;
}

.input-group.input-group_icon.input-group_icon-right .input-group__icon {
  right: 16px;
}
</style>
