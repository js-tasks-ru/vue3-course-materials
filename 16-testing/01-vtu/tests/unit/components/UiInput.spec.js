import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import UiInput from '@/components/UiInput';

// test('UiInput should be defined', () => {
//   expect(UiInput).toBeDefined();
// });

/*
  В этом файле используется mount, так как на этот момент на вебинаре мы ещё не узнали про shallowMount.
  Здесь подходит shallowMount.
 */

describe('UiInput', () => {
  it('should be defined', () => {
    expect(UiInput).toBeDefined();
  });

  // Ниже 4 не очень полезных теста. Добавлены для примера.
  // Их можно заменить на снапшот тесты
  // Но у них свои проблемы.
  it('should render input.input-group inside .form-control', async () => {
    const wrapper = mount(UiInput);
    const inputGroup = wrapper.find('.input-group');
    const formControl = inputGroup.find('input.form-control');
    expect(inputGroup.exists()).toBe(true);
    expect(formControl.exists()).toBe(true);
  });

  it('should have .form-control_rounded on input with rounded prop', async () => {
    const wrapper = mount(UiInput, { props: { rounded: true } });
    expect(wrapper.get('input').classes('form-control_rounded')).toBeTruthy();
  });

  it('should not have .form-control_rounded on input without rounded prop', async () => {
    const wrapper = mount(UiInput, { props: { rounded: false } });
    expect(wrapper.get('input').classes('form-control_rounded')).toBeFalsy();
  });

  it('should not have .form-control_rounded on input by default', async () => {
    const wrapper = mount(UiInput);
    expect(wrapper.get('input').classes('form-control_rounded')).toBeFalsy();
  });

  it('should toggle .form-control_rounded on input when rounded changes', async () => {
    const wrapper = mount(UiInput);
    const input = wrapper.get('input');

    await wrapper.setProps({ rounded: true });
    // await nextTick();
    expect(input.classes('form-control_rounded')).toBeTruthy();
  });

  it('should set input value from modelValue', async () => {
    const VALUE = 'TEST_VALUE';
    const wrapper = mount(UiInput, { props: { modelValue: VALUE } });
    const inputWrapper = wrapper.get('input');
    const input = inputWrapper.element;
    expect(input.value).toBe(VALUE);
  });

  it('should update input value on modelValue update', async () => {
    const NEW_VALUE = 'NEW_TEST_VALUE';
    const wrapper = mount(UiInput, { props: { modelValue: '' } });
    const inputWrapper = wrapper.get('input');
    // wrapper.get('input').element.value = value;
    // await wrapper.get('input').trigger('input');
    // await wrapper.get('input').trigger('change');
    await wrapper.get('input').setValue(NEW_VALUE);

    const input = inputWrapper.element;
    expect(input.value).toBe(NEW_VALUE);
  });

  it.each(['password', 'num', 'range', 'date', 'time', 'email', 'tel'])('should render input type=(%s)', (type) => {
    const wrapper = mount(UiInput, { attrs: { type } });
    expect(wrapper.get('input').attributes('type')).toBe(type);
  });

  it('should pass attrs to input', () => {
    const ATTRS = { type: 'email', maxlength: '20', 'data-attr-test': 'test' };
    const wrapper = mount(UiInput, { attrs: ATTRS });
    expect(wrapper.get('input').attributes()).toMatchObject(ATTRS);
  });

  it('should not pass attrs to root', () => {
    const ATTRS = { type: 'email', maxlength: '20', 'data-attr-test': 'test' };
    const wrapper = mount(UiInput, { attrs: ATTRS });
    expect(wrapper.attributes()).not.toMatchObject(ATTRS);
  });

  it('should emit update:modelValue with inputted value on input', async () => {
    const VALUE = 'TEST_VALUE';
    const wrapper = mount(UiInput);
    await wrapper.get('input').setValue(VALUE);
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted('update:modelValue')[0]).toEqual([VALUE]);
  });

  it('should handle input event when input listener is set', async () => {
    const value = 'TEST_VALUE';
    const handler = jest.fn();
    const wrapper = mount(UiInput, {
      attrs: {
        onInput: (event) => handler(event.target.value),
      },
    });

    await wrapper.get('input').setValue(value);

    expect(handler).toHaveBeenCalled();
    expect(handler).toHaveBeenCalledWith(value);
  });

  it('should match snapshot', async () => {
    const wrapper = mount(UiInput);
    // .__app._container - костыль из Issue для Vue 3
    // https://github.com/eddyerburgh/jest-serializer-vue/issues/49
    // Пока не видел версии, в которой работает нормально
    expect(wrapper.__app._container).toMatchSnapshot();
  });
});
