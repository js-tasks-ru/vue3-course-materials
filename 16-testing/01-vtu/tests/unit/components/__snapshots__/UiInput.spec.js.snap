// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiInput should match snapshot 1`] = `
<div
  data-v-app=""
>
  <div
    class="input-group"
  >
    <!--v-if-->
    <input
      class="form-control"
    />
    <!-- Или вместо customModel: -->
    <!-- :value="modelValue"-->
    <!-- @input="$emit('update:modelValue', $event.target.value)"-->
  </div>
</div>
`;
